\section{Theory}

\subsection{Modal haplotype}

Phyloage uses Y-STR mutation counting to estimate the TMRCA
(Time to Most Recent Common Ancestor) for a group of persons.
This is not an easy task because STR markers jump back and
forth. If a marker has changed from one value to another and
back again we do not know if a mutation has taken place or
not. So for long time spans STR mutation counting does not
work and we get wrong TMRCA estimates. In many cases this
effect can be calculated and compensated as it has been done
in \cite{Kly09}. However this method has it's limits.

Phyloage uses another approach. It combines SNP based phylogenetic
trees and Y-STR counting. Because of the use of SNPs, the
ancestral positions on the phylogenetic tree are well known.
Phyloage tries to calculate the ancestral haplotype for each
ancestor and uses Y-STR mutation counting between those
haplotypes and the modern samples. The calculated haplotypes
are not necessarily identical to the real ancestral values.
So they are called modal haplotypes.

The calculation of a modal haplotype is not always possible.
Figure \ref{modal} illustrates the situation.

\begin{figure}[ht]
\centering
\includegraphics[width=8.14cm]{img/modal.png}
\caption{\label{modal} Calculation of an ancestor's mutational
value from a set of samples. The calculation is only possible
if we have at least three samples.}
\end{figure}

Because mutation rates are usually very slow we may assume
that a mutation occurs rarely. In this case the modal value
of a mutation is identical to the value that occurs most
often among the samples. The rare cases are the mutations.
If the time spans get very long this is no longer true. We
have no idea how many mutations have occurred between the
ancestral haplotype and the modern samples. The mutational
values are just random. If we use them to calculate the modal
haplotype, the result will be different from the real ancestral
values.

So for a valid calculation of a modal haplotype, two
conditions must be satisfied:

\begin{enumerate}
\item We must have at least three samples and two of them must
	have the same mutational value.
\item The time span between two ancestral haplotypes must be
	small compared to the average time in which a mutation occurs.
\end{enumerate}


\subsection{Maximum parsimony method}

As default Phyloage uses an algorithm that tries to satisfy
the maximum parsimony criterion \cite{Wiki-Maximum_parsimony}.
This means that the tree that contains the least amount of
mutations is considered best.

Although this is very intuitive and widely used, it is not
always true or easy to calculate because

\begin{enumerate}
\item In many cases there is no best tree, but a number of
	viable solutions. As an easy example consider two persons
	with two different values for the same marker. Both values
	are valid solutions for the modal haplotype.
\item Long time single lineages may introduce a set of completely
	random values, because it is likely that marker's have changed
	several times. The criterion of maximum parsimony can be
	misleading \cite{Fels78}.
\item The calculation of the most parsimonious tree can require
	enormous amounts of computing time \cite{Wiki-Maximum_parsimony}.
	In such cases we prefer a good solution over the best one possible.
\end{enumerate}

\begin{figure}[ht]
\centering
\includegraphics[width=13cm]{img/parsimony.png}
\caption{\label{parsimony} An algorithm that satisfies the
criterion of maximum parsimony. The tree is calculated backwards
until a unique modal value is found. After that the missing
values are recalculated.}
\end{figure}

Figure \ref{parsimony} shows how the most parsimonious solution
is derived for a very simple phylogenetic tree. In this 
example the modal values are calculated backwards in time until
a valid solution is found. After that the tree is recalculated
from the past to the present to find modal values for nodes that
did not have a viable solution before.

In reality the situation is more complicated because in many
cases there simply is no valid solution for a certain marker.
The algorithm that is used by Phyloage works like this:

\begin{enumerate}
\item If possible, modal values that strictly satisfy the
	maximum parsimony criterion are calculated. These are
	the easy cases.
\item The rest of the modal values are calculated by using
	averages and real numbers. This solution should come
	close to the maximum parsimony criterion, at least for
	many samples. In reality this solution is not possible
	because real marker values are restricted to whole numbers.
\item The modal values are mapped to their nearest neighbors
	from the set of real marker values. Still some markers
	can not be calculated because they do not have a unique
	nearest neighbor.
\item All markers that could not be calculated in the top
	node are forced to a real world neighbor value. After
	that the whole tree is recalculated and all markers are
	forced to a real world value.
\end{enumerate}

The algorithm does not necessarily find the best solution
possible, but it should find a good one. For more information
please consult the program's source code documentation
\cite{PhyloageSourceDoc} directly.


\subsection{Phylofriend method}

Phyloage also provides an alternative method to calculate
the modal haplotypes. It is very simple and uses the same
algorithm as the Phylofriend \cite{Phylofriend} program.

\begin{figure}[ht]
\centering
\includegraphics[width=9.5cm]{img/phylofriend.png}
\caption{\label{phylofriend} The Phylofriend method to
determine a modal haplotype is very simple. If a modal
value can't be calculated, it just guesses and continues
the calculation.}
\end{figure}

Figure \ref{phylofriend} illustrates how Phylofriend calculates
a modal haplotype. Generally it chooses the most common value
as the ancestral value. If there is no most common value it
just guesses among the valid ones. This works very well if
a modal haplotype has many samples (at least three) downstream
of it's own. For just two samples the resulting modal haplotype
will be somewhere in between the two sample haplotypes.

This algorithm does not satisfy the maximum parsimony criterion,
but for sparse trees it works better than some might expect.
The reason is that a modal haplotype, that is calculated only
from samples, can not be influenced by older modal values afterwards,
thus eliminating the effect of distant single lineages.

What method to use depends highly on tree structure and
research goals. I am afraid that I can not give a general
recommendation. For sparsely populated trees both methods won't
work very well due to the difficulties involved.
For densely populated trees both methods, maximum parsimony
and Phylofriend, should yield similar results.


\subsection{\label{section_weights}Calculating weights for trees}

Each node of a tree connects several branches and each branch
yields it's own result when calculating the time to the common
ancestor. To calculate the age of a node, the ages of all substream
branches must be taken into account (see figure \ref{weights}).
The first versions of this
program used the average value of all branches. This is simple
but not correct because in this model a single lineage counts as
much as a densely populated branch.

\begin{figure}[ht]
\centering
\includegraphics[width=5.9cm]{img/weights.png}
\caption{\label{weights} A number N of samples with different age
estimates $a_1$ to $a_N$ to the common ancestor.}
\end{figure}

To get a better time estimate we must put weights to individual
branches so that single lineages count less and many lineages
count more. To calculate the weights we start with a weighted
time estimate:

\begin{equation}
a = \sum_{i = 1}^{N} w_i a_i    \label{timeestimate}
\end{equation}

\begin{tabular}{ll}
$a$: &  Average time estimate in arbitrary units\\
$a_i$: &  Age estimate of a single branch\\
$w_i$: &  Weight of a branch\\
N: &  Number of substream branches
\end{tabular}
\vspace{1em}

And because we are using a weighted average the sum of all
weights must be equal to 1:

\begin{equation}
\sum_{i = 1}^{N} w_i = 1  \label{normalization}
\end{equation}

How should the weights be calculated? A sensible condition
is that the relative error of our time estimate should be as small as
possible. According to the Gaussian error propagation law
the square of the standard deviation of $a$ is

\begin{equation}
\sigma_a^2 = \sum_{i = 1}^{N} \left(\frac{\partial a}{\partial a_i}
\right)^2 \sigma_i^2  \label{gauss}
\end{equation}

\begin{tabular}{ll}
$\sigma_a^2$: &  Standard deviation of the age estimate a\\
$a_i$: &  Age estimate of branch i\\
$\sigma_i$: &  Standard deviation of $a_i$\\
N: &  Number of substream branches
\end{tabular}
\vspace{1em}

Using equation \ref{timeestimate} we see that the derivatives
are the weights we are looking for:

\begin{equation}
\frac{\partial a}{\partial a_i} = w_i
\end{equation}

And thus we can write equation \ref{gauss} as

\begin{equation}
\sigma_a^2 = \sum_{i = 1}^{N} w_i^2 \sigma_i^2 \label{sigma_square}
\end{equation}

Our goal is to minimize the relative error. So $\sigma_a^2/a$
should be as small as possible. If $\sigma_a^2/a$ is constant
then all it's derivatives become zero and we have a minimum.

Equations \ref{timeestimate} and \ref{sigma_square} look very
similar and indeed if $\sigma_a^2/a$ should be constant we may
write

\begin{eqnarray}
\sigma_a^2 & = & c\ a \\
\Leftrightarrow\ \sum_{i = 1}^{N} w_i^2 \sigma_i^2 & = & c \sum_{i = 1}^{N} w_i a_i
\end{eqnarray}

\begin{tabular}{ll}
a: &  Age\\
$\sigma$: &  Standard deviation\\
w: & Weight\\
c: & Constant (yet unknown)\\
N: & Number of substream branches
\end{tabular}
\vspace{1em}

For symmetry reasons we can concentrate on a single term:

\begin{eqnarray}
w_i^2 \sigma_i^2 & = & c\ w_i a_i\\
\Leftrightarrow\ w_i & = & c\ \frac{a_i}{\sigma_i^2}
\end{eqnarray}

To calculate the constant $c$ we remember that the sum of all
weights must be one (equation \ref{normalization}) and we get 

\begin{equation}
w_i = \frac{1}{\sum_{j=1}^{N}\frac{a_j}{\sigma_j^2}} \ \frac{a_i}{\sigma_i^2}
\end{equation}

This is the formula for the weights. For real-world trees we also
need to calculate the standard deviations of all branches.
This is done in the next section.


\subsection{Calculating standard deviations}


\subsubsection*{Node with samples}

\begin{figure}[ht]
\centering
\includegraphics[width=5.9cm]{img/node_with_samples.png}
\caption{\label{node_with_samples} A number N of samples with different
numbers of mutations $m_1$ to $m_N$ to the common ancestral node.}
\end{figure}

It is common that a node is connected to several samples.
The distance from the node to each sample is a number of
mutations $m_i$. Figure \ref{node_with_samples} illustrates
the situation. In this case we can take advantage of the laws
of Gaussian statistics and the standard deviation for the
average number of mutation becomes:

\begin{eqnarray}
\sigma_{\bar{m}} & = & \sqrt{\frac{\bar{m}}{N}}\\
                 & = & \sqrt{\frac{1}{N^2} \sum_{i=1}^{N}m_i}\\
\Leftrightarrow\ \sigma_{\bar{m}}^2 & = & \frac{1}{N^2} \sum_{i=1}^{N}m_i
\end{eqnarray}

\begin{tabular}{ll}
$\sigma_{\bar{m}}$: &  Standard deviation of the average number of mutations\\
$\bar{m}$: & Average number of mutations to the common ancestor\\
$m_i$: &  Number of mutations in branch i\\
N: &  Number of branches
\end{tabular}
\vspace{1em}

This is also true for Poisson statistics but for Poisson
statistics the standard deviation becomes asymmetric to
the average number of mutations.


\subsubsection*{Node with branches}

\begin{figure}[ht]
\centering
\includegraphics[width=5.9cm]{img/node_with_branches.png}
\caption{\label{node_with_branches} A number N of branches
connected to a node. Each branch has it's own time estimate
$a_i$ and standard deviation $\sigma_i$.}
\end{figure}

A similar but not identical case is when a node is connected
to other nodes by several branches. For each branch a time
estimate and a standard deviation exists and we do not necessarily
know how they were determined (see figure \ref{node_with_branches}).

This is exactly the same situation as in section \ref{section_weights}
and we can write the age estimate as a weighted average:

\begin{equation}
a = \sum_{i = 1}^{N} w_i a_i
\end{equation}

\begin{tabular}{ll}
a: &  Total age estimate in arbitrary units\\
$a_i$: &  Age estimate of a single branch\\
w: &  Weight of a branch\\
N: &  Number of substream branches\\
$\sigma$: & Standard deviation
\end{tabular}
\vspace{1em}

Because we have already calculated this case we can use
equation \ref{sigma_square} to calculate the standard
deviation.

\begin{equation}
\sigma_a^2 = \sum_{i = 1}^{N} w_i^2 \sigma_i^2 
\end{equation}


\subsubsection*{Nodes in a row}

The last case is when three nodes are connected by
two branches in line (see figure \ref{nodes_in_row}).
Again each branch has it's own time estimate and
standard deviation. 

\begin{figure}[ht]
\centering
\includegraphics[width=7.7cm]{img/nodes_in_row.png}
\caption{\label{nodes_in_row} Two branches in a row.
Each branch has it's own time estimate
$a_i$ and standard deviation $\sigma_i$.}
\end{figure}

We get the total age estimate $a$ if we add the age estimates
of the two branches together:

\begin{equation}
a = a_1 + a_2
\end{equation}

The standard deviation is determined by the Gaussian error
propagation law:

\begin{eqnarray}
\sigma_a^2 & = & \left( \frac{\partial a}{\partial a_1} \right)^2 \sigma_{a_1}^2
+ \left( \frac{\partial a}{\partial a_2} \right)^2 \sigma_{a_2}^2\\
\Rightarrow\ \sigma_a^2 & = & \sigma_{a_1}^2 + \sigma_{a_2}^2
\end{eqnarray}

\begin{tabular}{ll}
$a_i$: &  Age estimate of a single branch in arbitrary units\\
$a$: & Total age\\
$\sigma$: &  Standard deviation
\end{tabular}
\vspace{1em}










